pkg/wasm/client.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/GoogleContainerTools/kpt/porch/pkg/oci"
+	"github.com/google/go-containerregistry/pkg/gcrane"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/stream"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+type Client struct {
+	*oci.Storage
+}
+
+func NewClient(cacheDir string) (*Client, error) {
+	store, err := oci.NewStorage(cacheDir)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{Storage: store}, nil
+}
+
+func (r *Client) PushWasm(ctx context.Context, wasmFile string, imageName string) error {
+	tag, err := name.NewTag(imageName)
+	if err != nil {
+		return fmt.Errorf("unable to parse tag %q: %v", imageName, err)
+	}
+
+	options := []remote.Option{
+		remote.WithAuthFromKeychain(gcrane.Keychain),
+		remote.WithContext(ctx),
+	}
+
+	rmt, err := remote.Get(tag, options...)
```

Comment:
I think you need to check the err != nil case as well, at least to check that the error is what you expect.

Reply:
Added it.